### ISSUE TITLE
Optimizing Multi-Tenant Database Connections in TypeORM with Custom PostgreSQL Handling

### ISSUE THREAD

**Issue by Jeremy-F** [typeorm/typeorm#10619]
### Feature Description

I am working on a multi-tenant application using TypeORM with PostgreSQL. The common approach of creating a separate DataSource for each tenant leads to inefficient database connection handling. This is because TypeORM's `PostgresDriver` creates a new connection pool each time a new DataSource is initialized. However, it's important to note that the pool itself does not manage the schema; this is handled by either the DataSource or the `PostgresDriver`, but never by the pool. I'm contemplating an alternative approach and would appreciate your insights.

**Problem:**
The standard practice of using a new DataSource for each tenant, while ensuring schema isolation, results in a high number of database connections, particularly under heavy load. This poses scalability challenges.

### The Solution

Would it be possible to add an optional parameter in `PostgresConnectionOptions` that would allow not creating a new pool, but using an existing one, if we find that the `connectionOptions` built in `PostgresDriver.createPool` are identical to another pool created previously?

# Alternative 

I propose using a "Custom PG version" approach. The idea is to override the pool creation logic in TypeORM's `PostgresDriver` to reuse a single connection pool across multiple tenant requests, thereby reducing the total number of connections.

**Implementation:**
1. **Global Pool Cache:** A cache to store pool instances.
2. **PoolBuilder Function:** A function to check the cache for an existing pool based on connection options, reusing it if available, or creating a new one otherwise.
3. **Custom PG Version:** Modifying the PostgreSQL driver to use `PoolBuilder` instead of the standard pool constructor.

**Code Example:**
```ts
const cache: { [key: string]: pg.Pool } = {};

function PoolBuilder(this: any, connectionOptions: any) {
  this.key = JSON.stringify(connectionOptions);
  if (!cache[this.key]) {
    cache[this.key] = new pg.Pool({ ...connectionOptions });
  }
  return cache[this.key];
}

const PG = { ...pg, Pool: PoolBuilder };

export const createSchemaDataSource = async (schema: string) => {
  const options = {
    ...ormconfig,
    schema,
    driver: PG,
  };
  const dataSource = new DataSource(options);
  return await dataSource.initialize();
};
```

**Conclusion:**
This approach is aimed at significantly reducing the number of open connections in a multi-tenant environment. However, I'm seeking feedback on potential risks or overlooked aspects of this strategy. The provided code is an example, and I am considering further enhancements, such as distributing connections across multiple pools to prevent system-wide failures.

**Are there any better alternatives or potential issues with handling this scenario? Your insights and suggestions would be highly valuable.**

Thank you in advance for your help!

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [X] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, and I know how to start.

**Comment by wdprice** [comment]
I'm facing a very similar use case right now, I am wondering you ever pursued this approach and had learnings that could be shared.

**Comment by Jeremy-F** [comment]
> I'm facing a very similar use case right now, I am wondering you ever pursued this approach and had learnings that could be shared.

No, unfortunately, we haven't implemented it in production yet. We have optimized queries and corrected certain queries that were mistakenly hitting the write instance due to the use of co.query. We've also tried to reduce the number of open connections as much as possible by optimizing parts of the code, such as caching.

However, we haven't mustered the courage to implement this in production. The main reason is fear—irrational, perhaps, but the inaction ensures that we don't make mistakes. Our primary concern is the potential mixing of client data, which would be far more catastrophic for us than a slow service. Additionally, this solution poses a problem in managing connection pools. With our primary solution of creating a connection (in the pg-pool sense) for each tenant, we were assured of having a minimum of 10 potential connections per tenant without needing to set a poolSize. With the proposed solution, we would have had to spend considerable time increasing the pool size or finding another way to manage it more effectively.

To be completely frank, I haven't been able to continue pushing TypeORM, and we are currently testing other ORMs like Drizzle, which should allow us to reuse the same connection without needing such hacks. And they are actively maintained.

On my part, I have thoroughly tested my proposed solution locally, and I believe it works.

**Comment by wdprice** [comment]
Thanks for the response @Jeremy-F. I have similar constraints and feelings about the TypeORM ecosystem, but am pretty heavily invested in it currently - a change would be a significant investment.

I found the approach used in this repo to be helpful, and it may be useful to you as well: https://github.com/Avallone-io/rls

We didn't integrate the package or fork it directly due to some requirements specific to our application that weren't compatible, but we used their approach of implementing a custom DataSource and custom QueryRunner which can dynamically wrap all queries with the tenant settings outlined in our RLS policies. This has proven a good way for us to continue using TypeORM 0.3.x and DataSources while enforcing our Postgres RLS effectively.
